### DISCUSSION TITLE
Execute MSSQL Stored Procedure with Parameters

### DISCUSSION THREAD

**Discussion by kevinr432** [sqlpage/SQLPage#655] [Q&A] [answered]
II wanted to call an MSSQL stored procedure from sqlpage.     To test, I created a very simple stored procedure with one input parameter and a simple out.   It works when called from SQL Enterprise manager.   But I cannot figure out how to call it from sql page.   Does anyone have any ideas?

Here is how I call it from SQL Enterprise Manager
execute SampleStoredProcedure 'This is a test!'

Here is how I tried to call it from sql page:
Execute SampleStoredProcedure $InputMessage;

See attachment.
![StoreProcedure](https://github.com/user-attachments/assets/c1588276-55e3-4846-929b-7764d0ddd0a9)

Thank you,
Kevin Reagan

**Comment by lovasoa** [comment, marked as answer]
Hello, and thank you for reporting this !

The SQL parser inside SQLPage currently does not support stored procedure calls with parameters in sql server. I will report this upstream and we will try to have this implemented.

In the mean time, calling stored procedures without parameters, and [user-defined functions](https://learn.microsoft.com/en-us/training/modules/create-stored-procedures-table-valued-functions/6-create-inline) works, across all supported databases.

### In your example case, you can create a user-defined function:

```sql
-- Create the function
CREATE FUNCTION SampleUserDefinedFunction
(
    @input VARCHAR(100)
)
RETURNS VARCHAR(200)
AS
BEGIN
    RETURN 'Your input was: ' + @input
END;
```

[try online](https://sqliteonline.com/#sqltext=%23url-sqlite%3Ddb-mssql%0D%0A%23tab-name%3DMS%20SQL%0D%0A--%20Create%20the%20function%0ACREATE%20FUNCTION%20SampleUserDefinedFunction%0A(%0A%20%20%20%20%40input%20VARCHAR(100)%0A)%0ARETURNS%20VARCHAR(200)%0AAS%0ABEGIN%0A%20%20%20%20RETURN%20'Your%20input%20was%3A%20'%20%2B%20%40input%0AEND%3B%0A)

### and then use it in a select statement

```sql
select dbo.SampleUserDefinedFunction($InputMessage);
```

[try online](https://sqliteonline.com/#sqltext=%23url-sqlite%3Ddb-mssql%0D%0A%23tab-name%3DMS%20SQL.1%0D%0A%0Aselect%20dbo.SampleUserDefinedFunction('hello')%3B)

**Reply by lovasoa** [reply]
Reported to the datafusion sql parser project as https://github.com/apache/datafusion-sqlparser-rs/issues/1489

**Reply by lovasoa** [reply]
I proposed a fix to the our sql parser; now waiting to get it approved: https://github.com/apache/datafusion-sqlparser-rs/pull/1490

**Comment by kevinr432** [comment]
The function works perfectly.  I don't know why I didn't think to try that.    Thank you so much!

**Reply by kevinr432** [reply]
Ultimately, I wanted to update tables in the stored procedure.    It seems that functions cannot update tables, so I think I will have to wait until the ability to call stored procedures with parameters is available.

**Reply by lovasoa** [reply]
What you can do already is use a table to pass parameters to your function. This can also serve as an event log.

```sql
insert into events values ('param 1', 'param2');
execute process_events;
```

**Comment by kevinr432** [comment]
That is a very good idea.  I think I can work with that until the execute storedprocedure with parameters becomes available.

 

Ultimately I want the put all of my complex editing and validation logic in the stored procedure and return “Success” or “Failed for XXXX reasons”

**Comment by lovasoa** [comment]
My contribution was accepted in the sql parser. Support for EXECUTE with parameters in SQL server will be in the next version of SQLPage !

**Comment by kevinr432** [comment]
Awesome.  Thank you.

**Comment by lovasoa** [comment]
fixed in v0.31 :)

**Reply by fpbaggio** [reply]
How to Use it？ has Demo？ ths

**Reply by lovasoa** [reply]
You use it with the normal syntax for MsSQL stored procedure; the same you already do outside of sqlpage:

```sql
 execute SampleStoredProcedure 'This is a test!'
```
